
# 🧩 Titanic Dataset Exploration with pandas

🎯 **Goal:**  
In this notebook, you'll explore the Titanic dataset using **pandas**.  
You'll practice the most common pandas functions for data inspection, selection, filtering, cleaning, and analysis.

For each function in the list below:
1. Explain what it does (in your own words, in a Markdown cell).
2. Give at least **two examples** using the Titanic dataset.
3. Add a short comment about the output or why it’s useful.


In [43]:
import pandas as pd

# Load Titanic dataset
# (Make sure titanic.csv is in the same folder as this notebook)
df = pd.read_csv("titanic_dataset.csv")

# Show first few rows
df.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S


## 🧠 Step 1: Inspecting the Data

Functions to explore:
- df.head()
- df.tail()
- df.info()
- df.describe()
- df.shape
- df.columns


## df.head()

### 1. Descripción:  
Es una función que por defecto devuelve las primeras 5 filas del dataframe.   
Si se pasa un int por parámetro, devuelve las primeras n filas.   
Si se pasa un int negativo por parámetro, devolverá todas las filas excepto las n últimas.  

<br>

### 2. Ejemplos:

In [44]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [45]:
df.head(-91)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
795          796         0       2   
796          797         1       1   
797          798         1       3   
798          799         0       3   
799          800         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
795                                 Otter, Mr. Richard    male  39.0      0   
796                        Leader, Dr. Alice (Farnham)  female  49.0      0   
797                                   Osman, Mrs. Mara  female  31.0      0   
798                       Ibrahim Shawah, Mr. Yousseff    male  30.0      0   
799  Van Impe, Mrs. Jean Baptiste (Rosalie Paula Go...  female  30.0      1   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
795      0             28213  13.0000   NaN        S  
796      0             17465  25.9292   D17        S  
797      0            349244   8.6833   NaN        S  
798      0              2685   7.2292   NaN        C  
799      1            345773  24.1500   NaN        S  

[800 rows x 12 columns]

### 3. Utilidad

Permite de manera rápida ver la estructura de un DataFrame.  

Si se ordenan los datos, nos permite crear nuevos DataFrames filtrando por los n primeros registros.  

Por ejemplo, si los datos estuviesen ordenados por edad:

<br>

**50 pasajeros más jóvenes** 

```python
df_50_pasajeros_mas_jovenes = df. head(50)
``` 
<br>

**Todos los pasajeros salvo los 50 más ancianos**

```python 
df_sin_50_mas_ancianos = df.head(-50)
```

---


## df.tail()

### 1. Descripción:  
Es una función que por defecto devuelve las primeras 5 últimas filas del dataframe.   
Si se pasa un int por parámetro, devuelve las últimas n filas.   
Si se pasa un int negativo por parámetro, devolverá todas las filas excepto las n primeras.  

<br>

### 2. Ejemplos:

In [46]:
df.tail(3)

PassengerId  Survived  Pclass                                      Name  \
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [47]:
df.tail(-100)

PassengerId  Survived  Pclass                                      Name  \
100          101         0       3                   Petranec, Miss. Matilda   
101          102         0       3          Petroff, Mr. Pastcho ("Pentcho")   
102          103         0       1                 White, Mr. Richard Frasar   
103          104         0       3                Johansson, Mr. Gustaf Joel   
104          105         0       3            Gustafsson, Mr. Anders Vilhelm   
..           ...       ...     ...                                       ...   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
100  female  28.0      0      0      349245   7.8958   NaN        S  
101    male   NaN      0      0      349215   7.8958   NaN        S  
102    male  21.0      0      1       35281  77.2875   D26        S  
103    male  33.0      0      0        7540   8.6542   NaN        S  
104    male  37.0      2      0     3101276   7.9250   NaN        S  
..      ...   ...    ...    ...         ...      ...   ...      ...  
886    male  27.0      0      0      211536  13.0000   NaN        S  
887  female  19.0      0      0      112053  30.0000   B42        S  
888  female   NaN      1      2  W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0      111369  30.0000  C148        C  
890    male  32.0      0      0      370376   7.7500   NaN        Q  

[791 rows x 12 columns]

### 3. Utilidad

Permite de manera rápida ver la estructura de un DataFrame.  

Si se ordenan los datos, nos permite crear nuevos DataFrames filtrando por los n últimos registros.  

Por ejemplo, si los datos estuviesen ordenados por edad:

<br>

**50 pasajeros más ancianos** 

```python
df_50_pasajeros_mas_ancianos = df.tail(50)
``` 
<br>

**Todos los pasajeros salvo los 50 más jóvenes**

```python 
df_sin_50_mas_jovenes = df.tail(-50)
```

---

## df.info()

### 1. Descripción:  
Imprime un resumen conciso de un DataFrame. Incluyendo el índices, columnas, dtypes, cantidad de no nulos y uso de memoria.

<br>

### 2. Ejemplos:

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### 3. Utilidad

Permite ver de manera rápida la estructura de un DataFrame.

---

## df.describe()

### 1. Descripción:  
Genera estadísticas descriptivas del DataFrame.

<br>

### 2. Ejemplos:

In [49]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [50]:
df.describe(include='object')

Name   Sex  Ticket Cabin Embarked
count                       891   891     891   204      889
unique                      891     2     681   147        3
top     Braund, Mr. Owen Harris  male  347082    G6        S
freq                          1   577       7     4      644

### 3. Utilidad

La función .describe() de pandas permite obtener un resumen estadístico rápido de las columnas numéricas de un DataFrame.  
Identifica valores extremos, medias, medianas y percentiles.  
También puede aplicarse a columnas no numéricas con:
```python
df.describe(include='object')
```

Este incluirá:   
count (cantidad de valores)  
unique (número de valores únicos),   
top (el valor más frecuente)  
freq (la frecuencia del valor más frecuente)  

---

## df.shape

### 1. Descripción:  
Devuelve una tupla que representa las dimesiones del DataFrame

<br>

### 2. Ejemplos:

In [51]:
df.shape

(891, 12)

### 3. Utilidad

Para el dataset de los pasajeros del Titanic indicaría que se trata de una tabla de 13 columnas por 891 filas.

---

## df.columns()

### 1. Descripción:  
Devuelve todas las etiquetas de las columnas del DataFrame.

<br>

### 2. Ejemplos:

In [52]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### 3.Utilidad

Podría ser util si obtenemos la lista de etiquetas para iterarla en algún bucle.

---


## 🔍 Step 2: Selecting Data

Functions to explore:
- df["column"]
- df[["col1", "col2"]]
- df.loc[]
- df.iloc[]


## df["column"]

### 1. Descripción:  
Devuelve los registros de la columna especificada en formato **Serie**.  


### 2. Ejemplos:


In [53]:
df["Age"].head()



0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

### 3.Utilidad

Para ver solo aquella o aquellas columnas en las que queramos trabajar.

---

## df["col1", "col2"]

### 1. Descripción:  
Devuelve los registros de las columnas especificadas en formato **DataFrame**.  


### 2. Ejemplos:


In [54]:
df[["Sex", "Age", "Survived"]].head()

Sex   Age  Survived
0    male  22.0         0
1  female  38.0         1
2  female  26.0         1
3  female  35.0         1
4    male  35.0         0

### 3.Utilidad

Para crear un sub-DataFrame con varias columnas para análisis o visualización.

---

## df.loc[]

### 1. Descripción:    

Accede a un grupo de filas y columnas a través de sus etiquetas.


### 2. Ejemplos:

In [55]:
df.loc[(df['Age'] < 18) & (df['Survived'] == 0), ['Name', 'Age']]

Name   Age
7             Palsson, Master. Gosta Leonard   2.0
14      Vestrom, Miss. Hulda Amanda Adolfina  14.0
16                      Rice, Master. Eugene   2.0
24             Palsson, Miss. Torborg Danira   8.0
50                Panula, Master. Juha Niilo   7.0
59        Goodwin, Master. William Frederick  11.0
63                     Skoog, Master. Harald   4.0
71                Goodwin, Miss. Lillian Amy  16.0
86                    Ford, Mr. William Neal  16.0
111                     Zabour, Miss. Hileni  14.5
114                    Attalah, Miss. Malake  17.0
119        Andersson, Miss. Ellis Anna Maria   2.0
138                      Osen, Mr. Olaf Elon  16.0
147         Ford, Miss. Robina Maggie "Ruby"   9.0
163                          Calic, Mr. Jovo  17.0
164             Panula, Master. Eino Viljami   1.0
171                     Rice, Master. Arthur   4.0
182    Asplund, Master. Clarence Gustaf Hugo   9.0
205               Strom, Miss. Telma Matilda   2.0
266                Panula, Mr. Ernesti Arvid  16.0
278                       Rice, Master. Eric   7.0
282                de Pelsmaeker, Mr. Alfons  16.0
297             Allison, Miss. Helen Loraine   2.0
333          Vander Planke, Mr. Leo Edmondus  16.0
352                       Elias, Mr. Tannous  15.0
374               Palsson, Miss. Stina Viola   3.0
386          Goodwin, Master. Sidney Leonard   1.0
419                Van Impe, Miss. Catharina  10.0
433               Kallio, Mr. Nikolai Erland  17.0
480           Goodwin, Master. Harold Victor   9.0
500                         Calic, Mr. Petar  17.0
532                     Elias, Mr. Joseph Jr  17.0
541     Andersson, Miss. Ingeborg Constanzia   9.0
542        Andersson, Miss. Sigrid Elisabeth  11.0
574             Rush, Mr. Alfred George John  16.0
634                       Skoog, Miss. Mabel   9.0
642            Skoog, Miss. Margit Elizabeth   2.0
683              Goodwin, Mr. Charles Edward  14.0
686                 Panula, Mr. Jaako Arnold  14.0
721                Jensen, Mr. Svend Lauritz  17.0
731                 Hassan, Mr. Houssein G N  11.0
746              Abbott, Mr. Rossmore Edward  16.0
764                   Eklund, Mr. Hans Linus  16.0
787                Rice, Master. George Hugh   8.0
791                      Gaskell, Mr. Alfred  16.0
813       Andersson, Miss. Ebba Iris Alfrida   6.0
819             Skoog, Master. Karl Thorsten  10.0
824             Panula, Master. Urho Abraham   2.0
841                 Mudd, Mr. Thomas Charles  16.0
844                      Culumovic, Mr. Jeso  17.0
850  Andersson, Master. Sigvard Harald Elias   4.0
852                  Boulos, Miss. Nourelain   9.0

In [56]:
df.loc[df['Sex']=='female','Name']

1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
8      Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                    Nasser, Mrs. Nicholas (Adele Achem)
                             ...                        
880         Shelley, Mrs. William (Imanita Parrish Hall)
882                         Dahlberg, Miss. Gerda Ulrika
885                 Rice, Mrs. William (Margaret Norton)
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
Name: Name, Length: 314, dtype: object

### 3.Utilidad

En el primer ejemplo calculo un dataframe con los pasajeros fallecidos menores de edad, en las filas paso las condiciones y en las columnas imprimo nombre y edad, al imprimir varias columnas me devuelve un DataFrame.

En el segundo ejemplo imprimo los nombres de todas las mujeres del barco, al imprimir solo una columna me devuelve una Serie.

---

## df.iloc[]

### 1. Descripción:  
Permite localizar registros basándonos en los índices del DataFrame.


### 2. Ejemplos:


In [57]:
df.loc[0:150, :]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
146          147         1       3   
147          148         0       3   
148          149         0       2   
149          150         0       2   
150          151         0       2   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
146       Andersson, Mr. August Edvard ("Wennerstrom")    male  27.0      0   
147                   Ford, Miss. Robina Maggie "Ruby"  female   9.0      2   
148           Navratil, Mr. Michel ("Louis M Hoffman")    male  36.5      0   
149                  Byles, Rev. Thomas Roussel Davids    male  42.0      0   
150                         Bateman, Rev. Robert James    male  51.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
146      0            350043   7.7958   NaN        S  
147      2        W./C. 6608  34.3750   NaN        S  
148      2            230080  26.0000    F2        S  
149      0            244310  13.0000   NaN        S  
150      0       S.O.P. 1166  12.5250   NaN        S  

[151 rows x 12 columns]

In [58]:
df.iloc[:,0::2]

PassengerId  Pclass     Sex  SibSp            Ticket Cabin
0              1       3    male      1         A/5 21171   NaN
1              2       1  female      1          PC 17599   C85
2              3       3  female      0  STON/O2. 3101282   NaN
3              4       1  female      1            113803  C123
4              5       3    male      0            373450   NaN
..           ...     ...     ...    ...               ...   ...
886          887       2    male      0            211536   NaN
887          888       1  female      0            112053   B42
888          889       3  female      1        W./C. 6607   NaN
889          890       1    male      0            111369  C148
890          891       3    male      0            370376   NaN

[891 rows x 6 columns]

### 3.Utilidad
En la primera llamada cojo los primeros 150 registros mostrando todas las columnas del DataFrame.

En la segunda llamada cojo todos los registros mostrando las columnas impares

---


## 🔎 Step 3: Filtering Rows

Functions to explore:
- df[df["Age"] > 30]
- df.query("Sex == 'female' and Survived == 1")


In [59]:
#La función muestra el dataframe con todos los pasajeros que cumplan la condición de que tengan más de 30 años.
df[df["Age"] > 30]



PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
4              5         0       3   
6              7         0       1   
11            12         1       1   
..           ...       ...     ...   
873          874         0       3   
879          880         1       1   
881          882         0       3   
885          886         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
881                                 Markun, Mr. Johann    male  33.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
3        0    113803  53.1000  C123        S  
4        0    373450   8.0500   NaN        S  
6        0     17463  51.8625   E46        S  
11       0    113783  26.5500  C103        S  
..     ...       ...      ...   ...      ...  
873      0    345765   9.0000   NaN        S  
879      1     11767  83.1583   C50        C  
881      0    349257   7.8958   NaN        S  
885      5    382652  29.1250   NaN        Q  
890      0    370376   7.7500   NaN        Q  

[305 rows x 12 columns]

In [60]:
#La función muestra los 5 primeros registros de aquellas mujeres que hayan sobrevivido al accidente del Titanic.
#La función query funciona pasándole una expresión booleana y devuelve aquellos registros que cumplan la condición.
#Puede recibir varias condiciones booleanas y un parámetro inplace:bool que sirve para modificar el dataframe en vez de crear uno nuevo.
df.query("Sex == 'female' and Survived == 1").head()

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C


## 🧹 Step 4: Handling Missing Data

Functions to explore:
- df.isna()
- df.isna().sum()
- df.dropna()
- df.fillna()


In [61]:

# Example: Check missing values
df.isna().sum()

# Fill missing ages with median
df["Age"] = df["Age"].fillna(df["Age"].median())
df["Age"].head()


0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

## df.isna()

### 1. Descripción:    

Función que detecta valores faltantes.
Devuelve un objeto con el mismo tamaño indicando si los valores son NA (None o nunpy.Nan) y los mapea a True, mientras que todo lo demás lo mapea a False.

### 2. Ejemplos:

In [62]:
df.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False  False  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [63]:
df.iloc[0:100, :].isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0         False     False   False  False  False  False  False  False   False   
1         False     False   False  False  False  False  False  False   False   
2         False     False   False  False  False  False  False  False   False   
3         False     False   False  False  False  False  False  False   False   
4         False     False   False  False  False  False  False  False   False   
..          ...       ...     ...    ...    ...    ...    ...    ...     ...   
95        False     False   False  False  False  False  False  False   False   
96        False     False   False  False  False  False  False  False   False   
97        False     False   False  False  False  False  False  False   False   
98        False     False   False  False  False  False  False  False   False   
99        False     False   False  False  False  False  False  False   False   

     Fare  Cabin  Embarked  
0   False   True     False  
1   False  False     False  
2   False   True     False  
3   False  False     False  
4   False   True     False  
..    ...    ...       ...  
95  False   True     False  
96  False  False     False  
97  False  False     False  
98  False   True     False  
99  False   True     False  

[100 rows x 12 columns]

### 3.Utilidad

Una manera muy simple de detectar valores nulos.  
Podemos combinarlo con otros métodos para afinar más el DataFrame a evaluar.

---

## df.isna().sum

### 1. Descripción:    

Función que devuelve la suma de los valores nulos de cada columna de la tabla.

### 2. Ejemplos:

In [64]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 3.Utilidad

Nos permite hacer un Count de manera muy simple de los valores nulos por cada columna del DataFrame.

---

## df.dropna()

### 1. Descripción:    

Función que remueve los valores faltantes. Puede recibir varios parámetros.  
<br>
**axis**: 0 para remover filas que contengan valores faltantes, 1 para remover columnas con valores faltantes.  

**how**: any, si hay al menos un valor dropea, all, solo dropea si todos los valores son Na.  

**thresh**: int (opt), alternativa a how, se le pasa cuantos valores Na ha de tener para dropear.  

**subset**: se le pasa uno o varios labels que dropearán si se dropea algo de lo anterior.

**inplace**: bool, si modifica o crea uno nuevo (False default)
### 2. Ejemplos:

In [65]:
df.dropna()

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[202 rows x 12 columns]

In [66]:
df.dropna(axis='columns')

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare  
0        0         A/5 21171   7.2500  
1        0          PC 17599  71.2833  
2        0  STON/O2. 3101282   7.9250  
3        0            113803  53.1000  
4        0            373450   8.0500  
..     ...               ...      ...  
886      0            211536  13.0000  
887      0            112053  30.0000  
888      2        W./C. 6607  23.4500  
889      0            111369  30.0000  
890      0            370376   7.7500  

[891 rows x 10 columns]

### 3.Utilidad

Nos permite borrar aquello que esté afectado por nulos, tanto por filas, columnas, subsets...  

En la segunda función, al llamar al eje columnas desaparecen las columnas que contenían valores nulos.

---

## df.fillna()

### 1. Descripción:    

Rellena aquellos valors NA/NaN y los reemplaza por el valor indicado. Puede recibir varios parámetros para elegir la forma en que se rellenarán los datos.

**value:** será el valor por el que se sustituirán todos los NA/NaN. 

**method**: ffill propaga el último valor válido hacia el siguiente, bfill usa el siguiente valor válido para cubrir el que falta.

**axis**: permite seleccionar el eje sobre el que cubrir los valores faltantes. Por defecto es 0 (filas)

**inplace**: modifica (True) o crea (False, por defecto) un nuevo dataframe.

**limit**: se utiliza junto con method, el máximo número de valores a rellenar. Si se sobrepasa el número deja de rellenar valores.



### 2. Ejemplos:

In [67]:
df.fillna('No data available')

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare              Cabin Embarked  
0        0         A/5 21171   7.2500  No data available        S  
1        0          PC 17599  71.2833                C85        C  
2        0  STON/O2. 3101282   7.9250  No data available        S  
3        0            113803  53.1000               C123        S  
4        0            373450   8.0500  No data available        S  
..     ...               ...      ...                ...      ...  
886      0            211536  13.0000  No data available        S  
887      0            112053  30.0000                B42        S  
888      2        W./C. 6607  23.4500  No data available        S  
889      0            111369  30.0000               C148        C  
890      0            370376   7.7500  No data available        Q  

[891 rows x 12 columns]

In [68]:
df.fillna(method='bfill')

/tmp/ipykernel_20771/2831856154.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='bfill')


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   C85        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250  C123        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   E46        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   B42        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500  C148        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [69]:
df.fillna(method='ffill')

/tmp/ipykernel_20771/1193302488.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill')


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   C85        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500  C123        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   C50        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500  C148        Q  

[891 rows x 12 columns]

### 3.Utilidad

Permite rellenar valores nulos de manera cómoda y por diferentes métodos de propagación.

---


## 📊 Step 5: Grouping and Aggregating

Functions to explore:
- df.groupby("Sex")["Survived"].mean()
- df["Fare"].mean()
- df["Age"].median()


In [70]:
#La función groupby precisa que se le pase una o más columnas para realizar la agrupación, es lo que va entre paréntesis, 
#Se pueden hacer operaciones sobre una columna con métodos que contiene la función como media, suma, etc
#Esta función calcula la media de supervivientes por sexo.
df.groupby("Sex")["Survived"].mean()

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

In [71]:
#Esta función aplica el precio medio pagado según la clase del billete.
df.groupby("Pclass")["Fare"].mean()

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

In [72]:
#La media de lo que se ha pagado por billete
df["Fare"].mean()

np.float64(32.204207968574636)

In [73]:
#Calcula la edad mediana, es decir, 28 (el resultado), si ordenamos todos las edades por cada pasajero, sería la que está en el medio de todas ellas.
df["Age"].median()

28.0


## 📈 Step 6: Sorting and Counting

Functions to explore:
- df.sort_values("Age")
- df["Sex"].unique()
- df["Pclass"].value_counts()


In [74]:
#Ordena los valores de la columna que se le pase de menor a mayor
#Se le puede pasar ascending=False para ordenar de mayor a menor.
#Con na_position = first/last los NaNs irán al principio o al final
df.sort_values("Age").head()



PassengerId  Survived  Pclass                             Name     Sex  \
803          804         1       3  Thomas, Master. Assad Alexander    male   
755          756         1       2        Hamalainen, Master. Viljo    male   
469          470         1       3    Baclini, Miss. Helene Barbara  female   
644          645         1       3           Baclini, Miss. Eugenie  female   
78            79         1       2    Caldwell, Master. Alden Gates    male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
803  0.42      0      1    2625   8.5167   NaN        C  
755  0.67      1      1  250649  14.5000   NaN        S  
469  0.75      2      1    2666  19.2583   NaN        C  
644  0.75      2      1    2666  19.2583   NaN        C  
78   0.83      0      2  248738  29.0000   NaN        S

In [75]:
#Pasa una lista con los valores que no se repiten de la columna especificada
df["Sex"].unique()

array(['male', 'female'], dtype=object)

In [76]:
#Devuelve una Serie con la frecuencia con la que se repite cada valor
#En este caso indica cuantos pasajeros hubo de casa clase

df["Pclass"].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [77]:
#Se puede pasar por parámetro normalize=True para que devuelva la proporción en vez de frecuencias.
#También que cuente los nulos si los hubiese con dropna="False" y ordenar los valores con ascending=True/False
df["Pclass"].value_counts(normalize=True, ascending=True)

Pclass
2    0.206510
1    0.242424
3    0.551066
Name: proportion, dtype: float64


## ⚙️ Step 7: Creating or Modifying Columns

Functions to explore:
- df.assign()
- df.apply()
- df["new_col"] = ...
- pd.concat()
- pd.merge()


## df.assign()

### 1. Descripción:    

Asigna una nueva columna al dataframe


### 2. Ejemplos:

In [78]:
df.assign(family_size = df['SibSp'] + df['Parch'] +1).head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  family_size  
0      0         A/5 21171   7.2500   NaN        S            2  
1      0          PC 17599  71.2833   C85        C            2  
2      0  STON/O2. 3101282   7.9250   NaN        S            1  
3      0            113803  53.1000  C123        S            2  
4      0            373450   8.0500   NaN        S            1

### 3.Utilidad

Permite de una manera siencilla crear columnas calculadas. En el ejemplo calculamos con cuántos familiares viaja cada pasajero, incluyéndose a si mismo.

---

## df.apply()

### 1. Descripción:    

Aplica una función a lo largo de todo un eje del DataFrame.

Recibe una función como parámetro.

**axis**: 0 o "index" para aplicar la funcion a cada columna; 1 o "columns" la aplica a cada fila

**raw**: False (por defecto, pasa las columnas o filas como Series a la función), True (la función recibe ndaray objetos, útil para aplicar NumPy reduction)

### 2. Ejemplos:

In [79]:
df["Age"].apply(lambda x: x*2)

0      44.0
1      76.0
2      52.0
3      70.0
4      70.0
       ... 
886    54.0
887    38.0
888    56.0
889    52.0
890    64.0
Name: Age, Length: 891, dtype: float64

### 3.Utilidad

Pertmite aplicar funciones de manera simple por fila o columna. Funciona bien con Numpy


---

## df['new_col'] =...

### 1. Descripción:    

Crea una nueva columna o reemplaza una existente definiendo los nuevos valores


### 2. Ejemplos:

In [80]:

# Example: Create new column
df["Fare_per_Age"] = df["Fare"] / df["Age"]
df[["Age", "Fare", "Fare_per_Age"]].head()


Age     Fare  Fare_per_Age
0  22.0   7.2500      0.329545
1  38.0  71.2833      1.875876
2  26.0   7.9250      0.304808
3  35.0  53.1000      1.517143
4  35.0   8.0500      0.230000

### 3.Utilidad

Simple y directo.  

En el ejemplo se crea una nueva columna con los valores de Fare/Age de cada fila.


---

## pd.concat

### 1. Descripción:    

Concatena objetos de panda sobre un eje en particular, debe recibir comom parámetros DataFrames o Series.

axis= 0 concatena por filas (defecto), 1 concatena por columnas

join = 'outer' (defecto) se toma la unión de índices/columnas, 'inner' toma la interseección de índices/ columnas

ignore_index = True reinicia índices, False (defecto) conserva índices originales.




### 2. Ejemplos:

In [81]:
df1 = df[['Name', 'Age']].head(5)
df2 = df[['Survived', 'Pclass']].head(5)

df_concat = pd.concat([df1, df2], axis=0, ignore_index=True)
df_concat

Name   Age  Survived  Pclass
0                            Braund, Mr. Owen Harris  22.0       NaN     NaN
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0       NaN     NaN
2                             Heikkinen, Miss. Laina  26.0       NaN     NaN
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0       NaN     NaN
4                           Allen, Mr. William Henry  35.0       NaN     NaN
5                                                NaN   NaN       0.0     3.0
6                                                NaN   NaN       1.0     1.0
7                                                NaN   NaN       1.0     3.0
8                                                NaN   NaN       1.0     1.0
9                                                NaN   NaN       0.0     3.0

In [82]:
df1 = df[['Name', 'Age']].head(5)
df2 = df[['Survived', 'Pclass']].head(3)

df_concat = pd.concat([df1, df2], axis=1)
df_concat

Name   Age  Survived  Pclass
0                            Braund, Mr. Owen Harris  22.0       0.0     3.0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0       1.0     1.0
2                             Heikkinen, Miss. Laina  26.0       1.0     3.0
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0       NaN     NaN
4                           Allen, Mr. William Henry  35.0       NaN     NaN

### 3.Utilidad

Apila DataFrames de forma vertical u horizontal. Si no coincide alguna columna completa con NaN.



---

## pd.merge

### 1. Descripción:    

Mergea Dataframes o Series al estilo sql


### 2. Ejemplos:

In [84]:
df_passengers = df[['PassengerId', 'Name', 'Pclass']].head(5)
df_survival = df[['PassengerId', 'Survived']].head(5)
print(df_passengers)
print(df_survival)


   PassengerId                                               Name  Pclass
0            1                            Braund, Mr. Owen Harris       3
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...       1
2            3                             Heikkinen, Miss. Laina       3
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1
4            5                           Allen, Mr. William Henry       3
   PassengerId  Survived
0            1         0
1            2         1
2            3         1
3            4         1
4            5         0


In [85]:
df_merged = pd.merge(df_passengers, df_survival, on='PassengerId', how='inner')
df_merged

PassengerId                                               Name  Pclass  \
0            1                            Braund, Mr. Owen Harris       3   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...       1   
2            3                             Heikkinen, Miss. Laina       3   
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1   
4            5                           Allen, Mr. William Henry       3   

   Survived  
0         0  
1         1  
2         1  
3         1  
4         0

In [86]:
df1 = df[['PassengerId', 'Name']].head(5)
df2 = df[['PassengerId', 'Survived']].head(3)  # solo 3 filas

print(df1)
print(df2)  

   PassengerId                                               Name
0            1                            Braund, Mr. Owen Harris
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...
2            3                             Heikkinen, Miss. Laina
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4            5                           Allen, Mr. William Henry
   PassengerId  Survived
0            1         0
1            2         1
2            3         1


In [87]:
df_merged_outer = pd.merge(df1, df2, on='PassengerId', how='outer')
df_merged_outer

PassengerId                                               Name  Survived
0            1                            Braund, Mr. Owen Harris       0.0
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...       1.0
2            3                             Heikkinen, Miss. Laina       1.0
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)       NaN
4            5                           Allen, Mr. William Henry       NaN

### 3.Utilidad

Te traes las funcionalidades de los joins de SQL de manera muy sencilla.  
Permite combinar información relacionada de diferentes fuentes de datos.


---


## 💾 Step 8: Exporting Data

Function to explore:
- df.to_csv("output.csv", index=False)


In [ ]:

#Una vez realizadas las operaciones con el Dataframe y las funciones de Pandas, se guarda el Dataframe resultante en un archivo CSV.
#Index=False para no escribir los ínidces en el archivo
#Este ejemplo devuelve el DataFrame original más la columna añadida en el apartado anterior con el ejemplo de df[newcol]=...

df.to_csv("titanic_cleaned.csv", index=False)



## 🧩 Step 9: Summary

Reflect on what you learned:
- Which functions were most useful?
- What insights did you gain from the Titanic dataset?
